### PR TITLE
First tests for the genesis implementation

### PR DESCRIPTION
**Pull request by facundominguez**
This PR contributes three tests that rehearse the genesis implementation in different ways.

Please, note that this PR targets the branch of #536. 

All the tests check that the immutable tip of the node under test ends up in the honest chain, at a block that is recent enough from the tip of the honest chain. Recent enough is defined as within s+d+1 slots of the tip of the honest chain, where s is the genesis window and d is the maximum Praos delay.

The tests differ in the inputs that they rehearse:
* "serve adversarial branches": tests schedules that provide as much time as necessary for the node under test to reach the expected immutable tip.
* "stalling leashing attack": tests schedules where adversarial peers sometimes omit sending random blocks or headers, which would cause the node under test to delay 
* "timed leashing attack": tests schedules where the schedules are executed up to a given time bound by which the node under test is expected to reach a recent immutable tip.

All tests are expected to fail at the moment since they depend on having a working genesis implementation.

Contents of the PR:
* The new genesis tests in commits https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/6ab827720f1aeba1a7915a3bd4580658c7a06828 and https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/01641f77c04807866d7bfea556a13f3db1b0065c
* Improvements to the peer simulator and the schedule generator in commits https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/065614cbebbdfa9bed158d590c56e0902b8eee58, https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/343695cde9eb7f7119042ce0dbf05a97400e8ec5, https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/e7ef3833c299b91415a4694fc61e7334ef3b036a, and https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/3714105301d6b485149f4325689041d7672b79fc
* Improvements to the schema generator in https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/441092ab7ebdfe58a76ab014e71612d18aa585e7 and https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/692c099597e5dd0ea320210cbf09824b163f7147
* Improvements to peer simulator tests in https://github.com/IntersectMBO/ouroboros-consensus/pull/840/commits/e3c0f5e7060752f6147a95896486b57e3475d510

### CODE REVIEW

**Comment by Niols** [comment]
Rebased on top of more recent `genesis/point-scheduler-squashed` (f02826f8d1f5a13390edd457225cba20c1235814) itself rebased on top of more recent main (706c4b162072a4ccf4d484007130f080f5c1a183).

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Setup.hs
Comment (dnadales) [inline ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Setup.hs:49]: Shouldn't this be a `TODO`? We could leave `FIXME` For things that *have* to be fixed before merging something.

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests.hs
Comment (dnadales) [inline ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests.hs:10]: Minor: we use both leading and trailing commas. Shouldn't we stick to one style? (we prefer leading commas).

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/LongRangeAttack.hs
Comment (dnadales) [inline ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/LongRangeAttack.hs:44]: Perhaps we should explain why `1` was chosen here.

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs
Comment (dnadales) [inline ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs:11]: Would it make sense to include a general explanation of what the `tests` in this module are testing?

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs
Comment (dnadales) [inline ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs:145]: I don't know if `rehearsed` means any of these:

```suggestion
-- A leashing attack would be exercised by a point schedule meeting either of
```

```suggestion
-- A leashing attack would be simulated by a point schedule meeting either of
```

```suggestion
-- A leashing attack would be modelled by a point schedule meeting either of
```

**Review by dnadales — APPROVED** [review]
I left some comments. Feel free to merge this PR. Thanks!

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests.hs
Comment (Niols) [reply ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests.hs:10]: No particular opinion on our side. Do you want us to change all the commas to leading wherever possible? Is there any tool that could take care of this (and/or other code styles you would want to enforce) automatically?

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Setup.hs
Comment (Niols) [reply ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Setup.hs]: Done in d96aee77387eccb196522b4e5b3df1b61518440c.

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/LongRangeAttack.hs
Comment (Niols) [reply ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/LongRangeAttack.hs:45]: Done in d96aee77387eccb196522b4e5b3df1b61518440c.

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs
Comment (Niols) [reply ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs:16]: I gave it a try in d96aee77387eccb196522b4e5b3df1b61518440c.

File: ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs
Comment (Niols) [reply ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs]: I went for “successfully conducted” in d96aee77387eccb196522b4e5b3df1b61518440c.

**Review by Niols — COMMENTED** [review]
I tried to address your comments in d96aee77387eccb196522b4e5b3df1b61518440c. Do tell me if that can be improved. Otherwise I suppose I'll merge this PR after https://github.com/IntersectMBO/ouroboros-consensus/pull/536.

**Comment by facundominguez** [comment]
Rebased after merging #536 to `main`.

**Pull request merged**
